### PR TITLE
PWN-7118 - swap info dialogs

### PR DESCRIPTION
**Pull request by NikolaiEmelianov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7118

## Description of Work

swap info dialogs

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/info/SwapInfoBottomSheet.kt [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+            .launchIn(lifecycleScope)
+    }
+
+    private suspend fun handleFeatureState(state: SwapState, tokens: List<SwapTokenModel>): Flow<List<AnyCellItem>> {
+        return when (state) {
+            SwapState.InitialLoading,
+            is SwapState.LoadingRoutes,
+            is SwapState.TokenAZero -> flowOf(mapper.mapEmptyLiquidityFee())
+            is SwapState.SwapException -> handleFeatureState(state.previousFeatureState, tokens)
+            is SwapState.SwapLoaded -> {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/info/SwapInfoBottomSheet.kt:145]: looks like we need a presenter here..

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/info/SwapInfoBottomSheet.kt [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+                        loadingCell
+                    }
+                    val fullUiList = mapper.mapEmptyLiquidityFee().plus(feeCells)
+                    emit(fullUiList)
+
+                    rateLoaderList.merge()
+                        .collect {
+                            val marketInfo = it.first
+                            val loadingCell = it.second
+                            val rateLoaderState = it.third
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/info/SwapInfoBottomSheet.kt:162]: maybe put it like (marketInfo, loadingCell, rateLoaderState)?

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/info/SwapInfoMapper.kt [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                text = TextContainer(R.string.swap_info_details_account_fee_subtitle)
+            )
+        )
+    }
+
+    fun mapLoadingLiquidityFee(
+        allTokens: List<SwapTokenModel>,
+        route: JupiterSwapRoute? = null,
+    ): List<AnyCellItem> = buildList {
+        addAll(mapEmptyLiquidityFee())
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/info/SwapInfoMapper.kt:61]: same this += mapEmptyLiquidityFee()

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/info/SwapInfoMapper.kt [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        return when (state) {
+            SwapRateLoaderState.Empty -> oldCell
+            is SwapRateLoaderState.NoRateAvailable,
+            is SwapRateLoaderState.Error -> oldCell.copy(rightSideCellModel = null)
+            SwapRateLoaderState.Loading -> oldCell.copy(rightSideCellModel = rightSideSkeleton())
+            is SwapRateLoaderState.Loaded -> {
+                val rate = state.rate
+                val token = state.token
+                val feeInUsd = marketInfo.lpFee.amountInLamports
+                    .fromLamports(token.decimals).multiply(rate).formatFiat()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/info/SwapInfoMapper.kt:165]:
```suggestion
                val feeInUsd = marketInfo.lpFee.amountInLamports
                    .fromLamports(token.decimals)
                    .multiply(rate)
                    .formatFiat()
```

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/settings/presenter/JupiterSwapSettingsPresenter.kt [hunk 5]
```diff
@@ -100,9 +101,27 @@ class JupiterSwapSettingsPresenter(
...
             }
+            is SwapSettingsPayload -> onDetailsClick(payload)
         }
     }
 
+    private fun onDetailsClick(settingsPayload: SwapSettingsPayload) {
+        when (settingsPayload) {
+            SwapSettingsPayload.ROUTE ->
+                if (canOpenDetails()) view?.showRouteDialog()
+            SwapSettingsPayload.NETWORK_FEE -> view?.showDetailsDialog(SwapInfoType.NETWORK_FEE)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/settings/presenter/JupiterSwapSettingsPresenter.kt:112]: better to have one style of wrapping when cases 
everything like 
```
-> 
    ....
```
or
`-> ...`

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/settings/presenter/SwapContentSettingsMapper.kt [hunk 6]
```diff
@@ -121,10 +121,9 @@ class SwapContentSettingsMapper(
         if (route == null) return emptyString()
         var result = ""
         route.marketInfos.forEachIndexed { index, marketInfo ->
-            result = if (index != route.marketInfos.lastIndex) {
-                result.plus(jupiterTokens.findTokenSymbolByMint(marketInfo.inputMint)).plus("→")
-            } else {
-                result.plus(jupiterTokens.findTokenSymbolByMint(marketInfo.outputMint))
+            result = result.plus(jupiterTokens.findTokenSymbolByMint(marketInfo.inputMint)).plus("→")
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/settings/presenter/SwapContentSettingsMapper.kt:124]: buildString to remove .plus

**Review by gslevinkov — APPROVED** [review]

**Review by 1sakov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/info/SwapInfoBottomSheet.kt (hunk 2 above)
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/info/SwapInfoBottomSheet.kt:161]: it doesn't work for Triple

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/info/SwapInfoBottomSheet.kt (hunk 1 above)
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/info/SwapInfoBottomSheet.kt:144]: refactoring later if add some local properties

**Pull request merged**